File: tasks/printinorder/PrintInOrder.java
package printinorder;

public class PrintInOrder {

    volatile private int currentThread = 1;

    public PrintInOrder() {
    }

    public synchronized void first(Runnable printFirst) throws InterruptedException {
        while (currentThread != 1) {
            wait();
        }
        printFirst.run();
        currentThread = 2;
        notifyAll();
    }

    public synchronized void second(Runnable printSecond) throws InterruptedException {
        while (currentThread != 2) {
            wait();
        }
        printSecond.run();
        currentThread = 3;
        notifyAll();
    }

    public synchronized void third(Runnable printThird) throws InterruptedException {
        while (currentThread != 3) {
            wait();
        }
        printThird.run();
        notifyAll();
    }
}

File: tasks/printinorder/PrintInOrder.kt
package printinorder

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock

class PrintInOrderAlternative {

    private val lock = ReentrantLock()
    val order: AtomicInteger = AtomicInteger(1)

    fun first(printFirst: () -> Unit) {
        while (order.get() != 1) {
            Thread.yield()
        }
        printFirst()
        order.getAndIncrement()
    }
    fun second(printSecond: () -> Unit) {
        while (order.get() != 2) {
            Thread.onSpinWait()
        }
        printSecond()
        order.getAndIncrement()
    }

    fun third(printThird: () -> Unit) {
        while (order.get() != 3) {
            Thread.yield()
        }
        printThird()
        order.getAndSet(1)
    }
}

fun main() {

    val printInOrder = PrintInOrderAlternative()

    val result = StringBuilder()

    val thread1 = Thread {
        printInOrder.first {
            result.append("first")
        }
    }

    val thread2 = Thread {
        printInOrder.second {
            result.append("second")
        }
    }

    val thread3 = Thread {
        printInOrder.third {
            result.append("third")
        }
    }

    thread3.start()
    thread2.start()
    thread1.start()

    thread1.join()
    thread2.join()
    thread3.join()

    println(result.toString())
}
